dynamicPermissions.groups.ps1

```diff
@@ -0,0 +1,456 @@
+    }
+    catch {
+        $ex = $PSItem
+        if ($($ex.Exception.GetType().FullName -eq "Microsoft.PowerShell.Commands.HttpResponseException") -or
+            $($ex.Exception.GetType().FullName -eq "System.Net.WebException")) {
+            $errorObj = Resolve-HelloIDError -ErrorObject $ex
+            $auditMessage = "Error querying groups. Error: $($errorObj.FriendlyMessage)"
+            Write-Warning "Error at Line [$($errorObj.ScriptLineNumber)]: $($errorObj.Line). Error: $($errorObj.ErrorDetails)"
+        }
+        else {
+            $auditMessage = "Error querying groups. Error: $($ex.Exception.Message)"
+            Write-Warning "Error at Line [$($ex.InvocationInfo.ScriptLineNumber)]: $($ex.InvocationInfo.Line). Error: $($ex.Exception.Message)"
+        }
+        $outputContext.AuditLogs.Add([PSCustomObject]@{
+                # Action  = "" # Optional
+                Message = $auditMessage
+                IsError = $true
+            })
+
+        # Throw terminal error
+        throw $auditMessage
+    }
+
+    $desiredPermissions = @{ }
+    if (-Not($actionContext.Operation -eq "revoke")) {
+        foreach ($contract in $personContext.Person.Contracts) {
+            Write-Verbose "Contract in condition: $($contract.Context.InConditions)"
+            if ($contract.Context.InConditions -OR ($actionContext.DryRun -eq $True)) {
+                # Example: department_<department externalId>
+                $groupName = "department_" + $contract.Department.ExternalId
+                $groupName = Remove-StringLatinCharacters $groupName
+
+                $permissionCorrelationValue = $groupName
+
+                Write-Verbose "Querying group where [$($permissionCorrelationField)] = [$($permissionCorrelationValue)]"
+
+                $correlatedResource = $null
+                $correlatedResource = $groupsGrouped["$($permissionCorrelationValue)"]
+
+                if (($correlatedResource | Measure-Object).count -eq 1) {
+                    # Add group to desired permissions with the id as key and the displayname as value (use id to avoid issues with name changes and for uniqueness)
+                    $desiredPermissions["$($correlatedResource.groupGuid)"] = $correlatedResource.name
+                }
+                elseif (($correlatedResource | Measure-Object).count -gt 1) {
+                    $auditMessage = "Multiple groups found where [$($permissionCorrelationField)] = [$($permissionCorrelationValue)]. Please correct this so the groups are unique."
+            
+                    $outputContext.AuditLogs.Add([PSCustomObject]@{
+                            # Action  = "" # Optional
+                            Message = $auditMessage
+                            IsError = $true
+                        })
+                
+                    # Throw terminal error
+                    throw $auditMessage
+                }
+            }
+        }
```

Comment:
When multiple groups or no groups are found, the script immediately throws an error. When a person has multiple incontidion contracts we should generate a error for each group that isn't found or is found multiple times.

Suggestion: Remove throw at found multiple groups and add the code below after the foreach

```powershell
        if ($outputContext.AuditLogs.IsError -contains $true) {
            throw "One or more errors occurred while looking up group(s)."
        }
```

Reply:
Agreed, has been changed.